src/printer/tvmscript_unified_printer.cc

```diff
@@ -0,0 +1,738 @@
+      PrintDoc(Downcast<AttrAccessDoc>(doc));
+    } else if (doc->IsInstance<IndexDocNode>()) {
+      PrintDoc(Downcast<IndexDoc>(doc));
+    } else if (doc->IsInstance<OperationDocNode>()) {
+      PrintDoc(Downcast<OperationDoc>(doc));
+    } else if (doc->IsInstance<CallDocNode>()) {
+      PrintDoc(Downcast<CallDoc>(doc));
+    } else if (doc->IsInstance<TupleDocNode>()) {
+      PrintDoc(Downcast<TupleDoc>(doc));
+    } else if (doc->IsInstance<SeqStmtDocNode>()) {
+      PrintDoc(Downcast<SeqStmtDoc>(doc));
+    } else if (doc->IsInstance<ScopeDocNode>()) {
+      PrintDoc(Downcast<ScopeDoc>(doc));
+    } else if (doc->IsInstance<ForDocNode>()) {
+      PrintDoc(Downcast<ForDoc>(doc));
+    } else if (doc->IsInstance<AssignDocNode>()) {
+      PrintDoc(Downcast<AssignDoc>(doc));
+    } else if (doc->IsInstance<ExprTypeDocNode>()) {
+      PrintDoc(Downcast<ExprTypeDoc>(doc));
+    } else if (doc->IsInstance<TypeCallDocNode>()) {
+      PrintDoc(Downcast<TypeCallDoc>(doc));
+    } else if (doc->IsInstance<FunctionDocNode>()) {
+      PrintDoc(Downcast<FunctionDoc>(doc));
+    } else if (doc->IsInstance<FunctionArgDocNode>()) {
+      PrintDoc(Downcast<FunctionArgDoc>(doc));
+    } else {
+      LOG(FATAL) << "Do not know how to print " << doc->GetTypeKey();
+    }
+  };
+
+  virtual void PrintDoc(const LiteralValueDoc& doc) = 0;
+  virtual void PrintDoc(const ConstDoc& doc) = 0;
+  virtual void PrintDoc(const IdentifierDoc& doc) = 0;
+  virtual void PrintDoc(const AttrAccessDoc& doc) = 0;
+  virtual void PrintDoc(const IndexDoc& doc) = 0;
+  virtual void PrintDoc(const OperationDoc& doc) = 0;
+  virtual void PrintDoc(const CallDoc& doc) = 0;
+  virtual void PrintDoc(const TupleDoc& doc) = 0;
+  virtual void PrintDoc(const SeqStmtDoc& doc) = 0;
+  virtual void PrintDoc(const ScopeDoc& doc) = 0;
+  virtual void PrintDoc(const ForDoc& doc) = 0;
+  virtual void PrintDoc(const AssignDoc& doc) = 0;
+  virtual void PrintDoc(const ExprTypeDoc& doc) = 0;
+  virtual void PrintDoc(const TypeCallDoc& doc) = 0;
+  virtual void PrintDoc(const FunctionDoc& doc) = 0;
+  virtual void PrintDoc(const FunctionArgDoc& doc) = 0;
+
+  using OutputStream = std::ostringstream;
+
+  OutputStream& NewLine() {
+    output_ << "\n" << std::string(indent_, ' ');
+    return output_;
+  }
+
+  OutputStream output_;
+  int indent_ = 0;
+};
+
+class PythonDocPrinter : public DocPrinter {
+ public:
+  PythonDocPrinter(String tir_prefix) : tir_prefix_(tir_prefix) {}
+
+ protected:
+  using DocPrinter::PrintDoc;
+
+  virtual void PrintDoc(const LiteralValueDoc& doc) override {
+    auto& value = doc->value;
+    if (value->IsInstance<FloatImmNode>() || value->IsInstance<IntImmNode>()) {
+      PrintNumberNode(Downcast<PrimExpr>(doc->value));
+    } else if (value->IsInstance<StringObj>()) {
+      PrintStringLiteral(Downcast<String>(value));
+    } else if (value->IsInstance<tir::StringImmNode>()) {
+      PrintStringLiteral(Downcast<StringImm>(value)->value);
+    }
```

Comment:
```suggestion
    } else if (const auto* v = value.as<tir::StringImmNode>()) {
      PrintStringLiteral(v->value);
    } else {
      LOG(FATAL) << "type blahblah: " << value->GetTypeKey();
      throw;
    }
```

Reply:
I found it's harder to apply this to other if predicates. For example,

```
  if (const auto* node = doc.as<LiteralValueDocNode>()) {
    PrintDoc(GetRef<LiteralValueDoc>(node));
```
still does type check twice. Is there a way to get ref from node pointer without type check? (since we already statically know node is a `LiteralValueDocNode*`, it should be safe to convert it to `LiteralValueDoc` without type checking)